File: concurrency/conditions/go/producerconsumer.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var mutex sync.Mutex = sync.Mutex{}

var EMPTY *sync.Cond = sync.NewCond(&mutex)
var FILL *sync.Cond = sync.NewCond(&mutex)

var MAX_SIZE int = 10

type List struct {
	buffer  []int
	fillIdx int
	useIdx  int
	count   int
}

var list List = List{
	buffer:  make([]int, MAX_SIZE),
	fillIdx: 0,
	useIdx:  0,
	count:   0,
}

func (list *List) put(i int) {
	list.buffer[list.fillIdx] = i
	list.fillIdx = (list.fillIdx + 1) % MAX_SIZE
	list.count++
}

func (list *List) get() int {
	tmp := list.buffer[list.useIdx]
	list.useIdx = (list.useIdx + 1) % MAX_SIZE
	list.count--
	return tmp
}

func (list *List) len() int {
	return list.count
}

type Producer struct {
	loops int
	name  string
}

func (p *Producer) run() {
	for i := 0; i < p.loops; i++ {
		mutex.Lock()

		for list.len() == MAX_SIZE {
			EMPTY.Wait()
		}
		list.put(i)
		fmt.Printf("produce: %d\n", i)
		FILL.Broadcast()
		mutex.Unlock()
	}
}

func InitProducer(loops int, name string) *Producer {
	p := Producer{
		loops: loops,
		name:  name,
	}
	return &p
}

type Consumer struct {
	name string
}

func InitConsumer(name string) *Consumer {
	consmer := Consumer{
		name: name,
	}
	return &consmer
}

func (c *Consumer) run() {
	for {
		mutex.Lock()
		for list.len() == 0 {
			FILL.Wait()
		}
		tmp := list.get()
		EMPTY.Signal()
		mutex.Unlock()
		fmt.Printf("%s consumes: %d\n", c.name, tmp)
	}
}

func main() {
	p := InitProducer(100, "Producer-1")
	c1 := InitConsumer("Consumer-1")
	c2 := InitConsumer("Consumer-2")

	go p.run()
	go c1.run()
	go c2.run()

	time.Sleep(5 * time.Second)
}

File: concurrency/conditions/java/ProducerConsumer.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ProducerConsumer {
    static final int MAX_SIZE = 10;
    static final Lock lock = new ReentrantLock();
    static Condition EPMTY = lock.newCondition();
    static Condition FILL = lock.newCondition();
    static List<Integer> list = new ArrayList<>(MAX_SIZE);

    static class Producer implements Runnable {
        final int loops;

        public Producer(int loops) {
            this.loops = loops;
        }

        @Override
        public void run() {
            for (int i = 0; i < loops; i++) {
                lock.lock();
                try {
                    while (list.size() == MAX_SIZE) {
                        EPMTY.await();
                    }
                    list.addLast(i);
                    FILL.signalAll();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    lock.unlock();
                }
            }
        }
    }

    static class Consumer implements Runnable {
        @Override
        public void run() {
            while (true) {
                lock.lock();
                try {
                    while (list.size() == 0) {
                        FILL.await();
                    }

                    int tmp = list.removeFirst();
                    EPMTY.signalAll();
                    System.out.printf("%s consume: %d\n", Thread.currentThread().getName(), tmp);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    lock.unlock();
                }
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        var producer = new Thread(new Producer(100), "Producer");
        var consumer1 = new Thread(new Consumer(), "Consmer-1");
        var consumer2 = new Thread(new Consumer(), "Consumer-2");

        producer.start();
        consumer1.start();
        consumer2.start();
        TimeUnit.SECONDS.sleep(5);
    }
}
